### ISSUE TITLE
Zipkin integration

### ISSUE THREAD

**Issue by rhart** [ratpack/ratpack#913]

**Comment by rhart** [comment]
Super cool. One approach that might work well is if you share your app and we can iterate there to start with. Start pulling out a module and then get a PR going. Do you fancy doing that?

**Comment by hyleung** [comment]
Yeah could give that a go. Tho' the repo where I did the work might not be ideal for that (it has a bunch of other stuff in there). But I can create a separate repo with a Ratpack app.

In any case, here's the way I hacked it into my particular app: https://github.com/hyleung/distributed-demo/blob/master/api-gateway/src/main/java/com/example/tracing/BraveTracingHandler.java

I'm using the [`brave-http`](https://github.com/openzipkin/brave/tree/master/brave-http), which has a bunch of adapter interfaces etc that allow you to use the interceptors - not 100% sure if I'm using the APIs correctly, but I am able to get traces. Mostly I had a look at `brave-jaxrs2` and followed a similar pattern.

**Comment by rhart** [comment]
I'm on hols at the mo and on my phone so will need to take a proper look when I'm back. As long as the spans being recorded are SR and SS then it should be on the right lines. 

We also need to look at recording CS and CR with the http client. 

I'm back next week so will be in touch soon

**Comment by rhart** [comment]
Taking a quick look at this we defo need to be using `HttpServerRequestAdapter` and not `ClientRequestAdapter `.  The server ones are for receiving requests and sending responses.  The client ones would be used with the http client, sending requests and receiving responses.

**Comment by rhart** [comment]
I think we need to lose these dependencies too
`compile 'org.glassfish.jersey.core:jersey-client:2.22.2'`
`compile 'com.github.kristofa:brave-jersey2:3.5.0'`

What are they used for?

**Comment by dnathanson** [comment]
You need to handle both ingress and egress points for HTTP messages.  So when receiving a message you need to look for trace headers in the request.  When sending a message (not sending a response to a request), if currently tracing, the trace headers that were found in the initial request need to be added as HTTP headers of requests that this service makes to another service in the chain.  So, I would expect you need to work with both the HttpServerRequestAdapter and the ClientRequestAdapter.

**Comment by rhart** [comment]
Exactly.  What I was saying is that in the example from @hyleung the tracing handler should be using the server adapters whereas it's currently using client ones.

When the example server gets extended to also make a client http request then we need to look at using the client adapters with Ratpack's http client.

**Comment by hyleung** [comment]
Hi all, I created a separate repo for this work (https://github.com/hyleung/ratpack-zipkin) and pretty much realized immediately that I was using the client request/response adapters instead of the server adapters. I've just been chipping away at this slowly...

I've removed the Jersey deps, which where there because on the original project I had been using Jersey prior to switching to Ratpack in a moment of rage. The new repo just has dependencies on:

``` 
 compile 'com.github.kristofa:brave-spancollector-scribe:3.5.0'
 compile 'com.github.kristofa:brave-http:3.5.0'
```

At the moment, I'm not pulling any annotations out of the request (other implementations do simple things like adding the request URI) - but I'll come back to that after I've got things working end-to-end. I'm wondering whether annotations are supposed to be something that people are able to add arbitrary data to.

**Comment by dnathanson** [comment]
You are correct. That is exactly what annotations are for.

**Comment by rhart** [comment]
@hyleung Great stuff!  

I think I would not depend on `brave-spancollector-scribe` and just use `LoggingSpanCollector ` provided by core as the default collector.  We should look this up from the registry to allow users to configure what collector they want and allow us to inject a test one for testing.  That would be a good next step.

**Comment by hyleung** [comment]
Alright, think I've got a first pass at SS/SR tracing over on my repo. Not having had a lot of experience with Ratpack, definitely could do with some tips on where there's a more "standard" pattern that could be used.

At the moment, it looks something like this:

```
Brave.Builder builder = new Brave.Builder("ratpack-demo");
RatpackServer.start(server ->
    server
        .serverConfig(config -> config.port(serverPort))
        .registry(registry(binding -> {
              binding
                  .module(ServerTracingModule.class)
                  .bind(HelloWorldHandler.class)
                  .bindInstance(ServerTracingModule
                      .config()
                      .withBrave(builder.build())
                      .withRequestAnnotations(request ->
                          Lists.newArrayList(KeyValueAnnotation.create("uri", request.getUri()))
                      )
                      .withResponseAnnotations(response ->
                          Lists.newArrayList(KeyValueAnnotation.create("foo", "bar"))
                      ));
            }
        ))
        .handlers(handler ->
            handler
                .get("hello", HelloWorldHandler.class)
                .all(ctx -> ctx.render("root")))
).start();
```

The `withRequestAnnotations`/`ResponseAnnotations` parts are how I'm getting the request/response annotations. Other Brave config (I'm using http://github.com/openzipkin/brave) is done using `Brave.Builder`.

**Comment by rhart** [comment]
This is looking great @hyleung.  I'll send you a PR with some suggested changes.

**Comment by rhart** [comment]
@hyleung I haven't looked through everything yet but PR sent with some suggestions so far.

We also need to look at implementing a registry backed `ServerClientAndLocalSpanState`.  If we don't override it then by default a thread local version will be used which is no good for us.

**Comment by rhart** [comment]
@hyleung while I remember, we also need to look at adding values into MDC.  SpanId, TraceId, AppName.  See `ratpack.logging.MDCInterceptor`.

**Comment by hyleung** [comment]
@rhart I added an issue in my repo as a reminder to look into the MDC values (https://github.com/hyleung/ratpack-zipkin/issues/5). Feel free to elaborate and/or add new issues to that repo.

**Comment by hyleung** [comment]
I'm going to start having a look at CS/CR with the client adapters. Does `HttpClient` have any hooks that I could use to inspect the request and response? Looking to see if there's any place I can use`ClientRequestInterceptor` and `ClientResponseInterceptor` (both from `brave-core`) to do the tracing.

**Comment by rhart** [comment]
I don't see anything @hyleung.  I think this warrants its own separate issue

**Comment by hyleung** [comment]
@rhart I've got something sort of working for CS/CR tracing, but it's pretty hacky. And i'm pretty sure it only works for GET and POST. I've created an issue on my repo here: https://github.com/hyleung/ratpack-zipkin/issues/7 .Could really do with some suggestions!

**Comment by tednaleid** [comment]
I'm not really familiar with the zipkin stuff, but I've been playing with ratpack executions recently and have found use in the [ExecInterceptor](https://ratpack.io/manual/current/api/ratpack/exec/ExecInterceptor.html) which lets you listen to the execution segments of a request as they happen.  The `Execution` is an instance of the ratpack registry and you can easily get access to the request and response there.  It might be a bit chatty as every execution segment is handed off to the interceptor, but you could omit all of the `BLOCKING` ones and the first `COMPUTE` one before starting to look at the request/response to see if they are fulfilled.

**Comment by rhart** [comment]
@hyleung will take a look soon as poss.

**Comment by rhart** [comment]
@tednaleid that's a good option but I think we need to provide something more convenient.  There will be other use cases for intercepting http client requests and responses.

**Comment by hyleung** [comment]
Maybe something to look at after (hopefully - I am still interested in moving this forward! Kinda took a detour to work on those interceptors) that I noticed in [go-kit](https://github.com/go-kit/kit). They originally had a module (or whatever you call it) specifically for supporting Zipkin, but it looks like they're deprecating it on favour of supporting it via OpenTracing (this would be the OpenTracing "driver" for Zipkin - https://github.com/openzipkin/zipkin-go-opentracing)

**Comment by rhart** [comment]
@hyleung what does supporting it via OpenTracing give you?

I will try and get to the client interceptors PR as soon as poss.

**Comment by hyleung** [comment]
@rhart supposedly, doing your tracing via OpenTracing allows you to plug in other tracing back-ends. With Go-kit, for example, there are "drivers" for AppDash (written in Go, open source), LightStep (I _think_ this one is commercial), as well as one for Zipkin. Anyhow, definitely will focus on getting Zipkin support first!

**Comment by yurishkuro** [comment]
@hyleung you're correct, integrating OpenTracing API allows the applications to make a choice which tracing library / backend they actually want to use. BTW, Jaeger (http://github.com/jaegertracing/jaeger) is a CNCF project that is OpenTracing compatible.

**Comment by johnrengelman** [comment]
FYI - HttpClient interceptors are now available in `master` - https://github.com/ratpack/ratpack/commit/b9bd8cf1807d5b12d5b389438813f485615f0e13
This should make integrating zipkin much easier as you can could add a handler that reads the incoming request and adds a new `HttpClient` with and outgoing intercept to add the headers.

**Comment by hyleung** [comment]
Oh wow, this is a blast from the past - yeah we (me, with *a lot* of help from other folks) actually went ahead and implemented a library that provides Zipkin support (https://github.com/hyleung/ratpack-zipkin), via https://github.com/openzipkin/brave.

@johnrengelman yeah I saw that HttpClient interceptors are now available. I tried taking a look at maybe reworking ratpack-zipkin to make use of the new API, but ran into a snag. IIRC this was because the way I had implemented the Zipkin support meant that I was trying to access  the context from a non-Ratpack managed thread, which was throwing a `UnmanagedThreadException`. I would like to come back to it eventually though as we had to shave a few yaks to get the client tracing working.

In any case, it's been so long I don't even what the state of the world is anymore with regard to Ratpack and Zipkin - is this issue still "a thing" (i.e. should it be closed or superseded by something else)? //cc @rhart

**Comment by beckje01** [comment]
At SmartThings we use the library heavily I think its worth closing this issue and adding a pointer in the manual to the external project unless there are any objections.

**Comment by rhart** [comment]
I have no objection as such but the intention was always to work something up externally and move it into the Ratpack repo when it was ready.  The main benefit being that we can ensure we don't break the integration with new releases.  Also, IMHO, I think tracing integration is something we should supply.

However, this all got held up with not reviewing PRs promptly and lack of time to spend on it.  I don't mind closing for now and revisiting when I have more bandwidth.

**Comment by beckje01** [comment]
@rhart tracing and zipkin in particular move pretty fast we can think about ways to trigger downstream tests for new builds if we want

**Comment by hyleung** [comment]
Originally, the idea _was_ to develop something that would eventually be brought back into the main Ratpack repo (the package names, etc. reflect this). After chipping away at this for a while, I agree with @beckje01. The Zipkin stuff moves pretty fast and at the very least, we'd want to be able to put out releases more frequently.

There are definitely some small tweaks that could be make to make things a bit easier though. For example, recently we wanted to be able to access the path binding so that we could add that info to the trace - but that's (as far as I could tell) accessible on the request, so we had to work around it.

Further down the road - just spitballing here - maybe at some point there could be more "abstract" tracing support in Ratpack, that would allow different tracing back-ends to be plugged in. T

**Comment by johnrengelman** [comment]
I'm fine with keeping it external, my point was that the code in `ratpack-zipkin` can collapse to something much simpler by utilizing the http client interceptors. Namely, you don't have to extend/replicate the `HttpClient` class anymore and you could perform the injection using the ratpack `Registry` and a `Handler` instead of relying on Guice binding annotations. This would offer much wider functionality to users (who aren't using Guice).

**Comment by beckje01** [comment]
@johnrengelman yep that work will probably happen soon.

**Comment by hyleung** [comment]
@johnrengelman ah i see. Yup absolutely agree that we should try to make use of the interceptors. I've added an issue so that it's not forgotten (https://github.com/hyleung/ratpack-zipkin/issues/44). 

Regarding Guice - it *is* being used to wire up a lot of stuff, so that might be a bit more work

**Comment by hyleung** [comment]
@johnrengelman I've started to look at reworking the http client in ratpack-zipkin, using the 1.6.0-SNAPSHOT. You mentioned above that I might be able to perform the injection using the ratpack `Registry` and a `Handler`? I noticed the `HttpClient.copyWith` stuff (which I think would work), but couldn't quite see how to use the `Registry` to inject intercepters. Can you point me in the right direction?

**Comment by johnrengelman** [comment]
The idea would be to add an `all` handler to the request chain, that gets the `HttpClient` from the `Registry` and calling `copyWith {}` on it to inject interceptors based on the incoming request and then calling `context.next(single(newHttpClient))` which injects the new `HttpClient` into the registry and continues the action.

so something like:
```
handlers {
  all {
    next(Registry.single(get(HttpClient).copyWith {
      requestIntercept { spec ->
        spec.headers { h -> h.add("SpanId", context.request.headers.get("SpanId")) }
      }
    }))
  }
  get { HttpClient client ->
    client.get("http://someotherservice.com").then {
      render "ok"
    }
  }
}
```

**Comment by johnrengelman** [comment]
This spec does the basic idea (But without reading from the incoming request) - https://github.com/ratpack/ratpack/blob/master/ratpack-core/src/test/groovy/ratpack/http/client/HttpClientInterceptorSpec.groovy#L84-L100

**Comment by hyleung** [comment]
@johnrengelman ahhh...cool! I think since I've already got something intercepting the incoming request  (for starting the Zipkin span) I might try injecting it at that point. Never occurred to me to try that, thanks!

**Comment by JohnRoesler** [comment]
Just discovered this issue! Anything I can help out with? I'm a pretty new developer, but excited to learn, and I've got a ratpack groovy api I'd like to add tracing to.

**Comment by hyleung** [comment]
@JohnRoesler You could the this a try: https://github.com/hyleung/ratpack-zipkin. The current latest version is [2.3.1](http://search.maven.org/#artifactdetails%7Ccom.github.hyleung%7Cratpack-zipkin%7C2.3.1%7Cjar) (just released w/Brave 4.16.1). I haven't tried it with a Groovy-based Ratpack API but I think it _should_ work. It does (currently) rely on `ratpack-guice` tho'...
